exporter/exporterhelper/internal/request/request.go

```diff
@@ -0,0 +1,69 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package request // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal/request"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/exporter/exporterhelper/request"
+)
+
+// Request is a wrapper around request.Request which adds context and a callback to be called when the request
+// is finished processing.
+type Request struct {
+	request.Request
+	ctx                          context.Context
+	onProcessingFinishedCallback func()
+}
+
+func New(ctx context.Context, req request.Request) *Request {
+	return &Request{
+		Request: req,
+		ctx:     ctx,
+	}
+}
+
+// Context returns the context.Context of the requests.
+func (req *Request) Context() context.Context {
+	return req.ctx
+}
+
+// SetContext updates the context.Context of the request.
+func (req *Request) SetContext(ctx context.Context) {
+	req.ctx = ctx
+}
+
+func (req *Request) OnError(err error) *Request {
+	if r, ok := req.Request.(request.ErrorHandler); ok {
+		return New(req.ctx, r.OnError(err))
+	}
+	return req
+}
+
+// Count returns a number of items in the request. If the request does not implement RequestItemsCounter
+// then 0 is returned.
+func (req *Request) Count() int {
+	if counter, ok := req.Request.(request.ItemsCounter); ok {
+		return counter.ItemsCount()
+	}
+	return 0
+}
+
+// OnProcessingFinished calls the optional callback function to handle cleanup after all processing is finished
+func (req *Request) OnProcessingFinished() {
```

Comment:
If we expose the queue interface, this probably can be exposed as an optional request interface as well. Then the queue `Produce` method can accept a public `Request` instead of the internal one